flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/reuse/ReplaceScanWithCalcShuttle.java

```diff
@@ -57,4 +61,12 @@ public RelNode visit(RelNode rel) {
 
         return super.visit(rel);
     }
+
+    private void visitDppSource(RelNode scan) {
+        // If scan is BatchPhysicalDynamicFilteringTableSourceScan,the input should be recursive
+        // first
```

Comment:
I think there needn't to strictly limit scan is dpp scan in the comments. But let's talk about sources with input, such as dpp sources